README.md
upload-service
==============

A File-Upload Web Application capable of dealing with *LARGE* files. The
backend is written in Perl using http://mojolicious.org, the frontend
leverages http://github.com/23/resumable.js

installation
------------
Use the classic ./configure; make install approach.

Once installed you will find a htdoc directory in your installation
target.

Copy this directory into your webserver and configure the webserver to
execute fcgi (fastcgi) scripts. Note that in the htdocs directory
there is also a .htaccess file.

Since the file-upload service is based on Mojolicious, you can use any of
the modes of operation supported by Mojolicious.

configuration
-------------

The upload service gets is configured via several environement variables.

Set US_ROOT to define where the uploaded files should end up

 US_ROOT=/path

The system has two modes of operatation. Single user mode and multi user mode

 US_SINGLEUSER=0|1

in single user mode the uploads go straight to $US_ROOT in multi user mode the
uploads go to $US_ROOT/$USERNAME/INBOX

in multi user mode the url for upload is expected to contain the $USERNAME
and the script must run as root (use sudo) so that it can suid to the
recipient of the files.

In single user mode you get

 http://your-site/script-url/

In multi-user mode you get

 http://your-site/script-url/USERNAME/

In multi user mode, the MOJO_TMPDIR must point to an empty directory
with mode 0777.

If you want to hand out 'individual' upload urls, you can enable tag mode

 US_TAGMODE=1

In tag mode, the upload url will look like this:

 http://your-site/script-url/TAG

the tag must be made up from [0-9a-z]+

Once the upload is complete the name of the uploaded file(s) will start with the given tag.

Enjoy!
tobi


CHANGES
2016-06-14 oetiker switched to new configure
2016-06-14 oetiker added US_TAGMODE


backend/lib/JfuPerl.pm
package JfuPerl;
use Mojo::Base 'Mojolicious';
use Mojo::Asset::File;
use Data::Dumper;
use Mojo::Util qw(hmac_sha1_sum b64_encode);
use POSIX qw(strftime);
# enable receiving uploads up to 1GB
$ENV{MOJO_MAX_MESSAGE_SIZE} = 1_073_741_824;

has root_dir => '/tmp/sandbox/';

sub startup {
    my $self = shift;
    my $me = $self;

    # properly figure your own path when running under fastcgi    
    $self->hook( before_dispatch => sub {
        my $self = shift;
        $self->res->headers->header('Access-Control-Allow-Origin' => '*');
        my $reqEnv = $self->req->env;
        my $uri = $reqEnv->{SCRIPT_URI} || $reqEnv->{REQUEST_URI};
        my $path_info = $reqEnv->{PATH_INFO};
        $uri =~ s|/?${path_info}$|/| if $path_info and $uri;
        $self->req->url->base(Mojo::URL->new($uri)) if $uri;
    });
    # session is valid for 1 day
    $self->secret('a093l;afslihj;akj;ljasdf.,mdfffdf');
    $self->sessions->cookie_name('uploader');
    $self->sessions->default_expiration(1*24*3600);

    my $r = $self->routes;

    # / (upload page)
    $r->get('/' => 'home');

    my $a = $r->under('/:user' => sub {
        my $self = shift;    
        my $root = $me->root_dir . '/'. $self->param('user') . '/INBOX';
        $self->stash(root=>$root);

        my $uid = getpwnam($self->param('user'));
        if (not $uid){
           $self->res->code(403);
           $self->render( text => 'unknown user');
           return;
        }
        # lets see if  we can do a bit of user switching
        if ($< != $uid ){
            # switch uid
            $> = $uid;
            if ($> != $uid){
                $self->res->code(403);
                $self->render( text => 'access denied: '.$!);
                return;
            }
        }
        
        if ( -l $root or not -w $root or not -d $root ) {
            $self->res->code(403);
            $self->render( text => 'no INBOX directory in sandbox');
            return;
        }

        my $sessionkey = $self->session('skey');
        if (not $sessionkey){
            my $newKey =hmac_sha1_sum(rand,time);
            $self->session(skey => $newKey);
        }
        $self->render_later;
    });

    # / (upload page)
    $a->get('/' => sub {
        my $self = shift;
                
        if ($self->req->url !~ m{/$}){
            $self->redirect_to($self->req->url.'/');
        }
        else {
            $self->render(template=>'uploadpage');
        }
    });

    # GET /upload (retrieves stored file list)
    $a->get('/upload' => sub {
        my $self = shift;
        my $sessionkey = $self->session('skey');
        my $root = $self->stash('root');
        my @list;
        for my $link (glob $root.'/.'.$sessionkey.'*'){
            next unless -l $link;
            my $dest = readlink $link;
            my $file = Mojo::Asset::File->new(path=>$root.'/'.$dest);
                        
            next unless $file->is_file;
            push @list, {
                name => $dest,
                size => $file->size,
                url => 'download/'.$dest,
                deleteUrl =>  'delete/'.$dest,
                deleteType => 'DELETE'
            };
        }
        return $self->render( json => { files => \@list } );
    });

    # POST /upload (push one or more files to app)
    $a->post('/upload' => sub {
        my $self    = shift;
        my @uploads = $self->req->upload('files[]');

        my @files;
        my $sessionkey = $self->session('skey');
        my $root = $self->stash('root');
        for my $upload (@uploads) {
            my $filename = $upload->filename;
            my $outfile = strftime("%Y-%m-%d_%H:%M:%S-$filename",localtime(time));
            $outfile =~ s{/}{_}g;
            if (symlink $outfile, $root. '/.'. $sessionkey . $outfile and  not -e $root. '/'.  $outfile ){
                push @files, {
                    name => $outfile,
                    size => $upload->size,
                    deleteUrl => 'delete/'.$outfile,
                    deleteType => 'DELETE',
                    url => 'download/'.$outfile
                };
                $upload->move_to( $root. '/'.  $outfile ) ;
            }
        }
        # return JSON list of uploads
        $self->render( json => { files => \@files } );
    });

    # /download/files/foo.txt
    $a->get('/download/#key' => sub {
        my $self = shift;        
        my $key  = $self->param('key');
        my $sessionkey = $self->session('skey');
        my $root = $self->stash('root');
        if (not -l $root .'/.'.$sessionkey.$key ){
            $self->res->code(403);
            $self->render( text => 'access denied');
            return;
        }
        my $file = Mojo::Asset::File->new(path=>$root . '/'. $key);
        if (not $file->is_file){
            $self->res->code(403);
            $self->render( text => 'access denied');
            return;
        }
        $self->render(
            data   => $file->slurp,
            format => 'application/octet-stream'
        );
    });
    
    # /delete/files/bar.tar.gz
    $a->delete('/delete/#key' => sub {
        my $self = shift;
        my $key  = $self->param('key');
        my $sessionkey = $self->session('skey');
        my $root = $self->stash('root');
        if (not -l $root .'/.'.$sessionkey.$key or -l $root ){
            $self->res->code(403);
            $self->render( text => 'access denied');
            return;
        }
        my $file = Mojo::Asset::File->new(path=>$root . '/'. $key);
        if (not $file->is_file){
            $self->res->code(403);
            $self->render( text => 'access denied');
            return;
        }
        unlink $file->path;
        unlink $root .'/.'.$sessionkey.$key;
        $self->render( json => 1 );
    });

}

1;


backend/public/js/main.js
/*
 * jQuery File Upload Plugin JS Example 8.9.0
 * https://github.com/blueimp/jQuery-File-Upload
 *
 * Copyright 2010, Sebastian Tschan
 * https://blueimp.net
 *
 * Licensed under the MIT license:
 * http://www.opensource.org/licenses/MIT
 */

/*jslint nomen: true, regexp: true */
/*global $, window, blueimp */

$(function () {
    'use strict';

    // Initialize the jQuery File Upload widget:
    $('#fileupload').fileupload({
        // Uncomment the following to send cross-domain cookies:
        //xhrFields: {withCredentials: true},
        url: 'upload'
    });

    // Enable iframe cross-domain access via redirect option:
    $('#fileupload').fileupload(
        'option',
        'redirect',
        window.location.href.replace(
            /\/[^\/]*$/,
            '/cors/result.html?%s'
        )
    );

    // Load existing files:
    $('#fileupload').addClass('fileupload-processing');
    $.ajax({
        // Uncomment the following to send cross-domain cookies:
        //xhrFields: {withCredentials: true},
        url: $('#fileupload').fileupload('option', 'url'),
        dataType: 'json',
        context: $('#fileupload')[0]
    }).always(function () {
        $(this).removeClass('fileupload-processing');
    }).done(function (result) {
        $(this).fileupload('option', 'done')
            .call(this, $.Event('done'), {result: result});
    });


});


lib/UploadService.pm
package UploadService;
use Mojo::Base 'Mojolicious';
use Mojo::Asset::File;
use Data::Dumper;
use Mojo::Util qw(hmac_sha1_sum b64_encode slurp);
use POSIX qw(strftime);
use Fcntl 'SEEK_SET';

our $VERSION = '0.0.0';

sub startup {
    my $self = shift;
    my $me = $self;

    # properly figure your own path when running under fastcgi
    $self->hook( before_dispatch => sub {
        my $self = shift;
        my $reqEnv = $self->req->env;
        my $uri = $reqEnv->{SCRIPT_URI} || $reqEnv->{REQUEST_URI};
        my $path_info = $reqEnv->{PATH_INFO};
        $uri =~ s|/?${path_info}$|/| if $path_info and $uri;
        $self->req->url->base(Mojo::URL->new($uri)) if $uri;
    });

    # session is valid for 1 day
    $self->secret(slurp($ENV{US_SECRET_FILE})) if $ENV{US_SECRET_FILE} and -r $ENV{US_SECRET_FILE};
    $self->sessions->cookie_name('uploader');
    $self->sessions->default_expiration(1*24*3600);

    my $r = $self->routes;

    my $b;
    # some intial checks
    if ($ENV{US_SINGLEUSER}){
       $b = $r->under(sub {
            my $self = shift;
            $self->stash(user=>scalar getpwuid($>));
            my $root = $ENV{MOJO_TMPDIR} = $ENV{US_ROOT};
            $self->stash(root => $root);
            if ( not -w $root or not -d $root ) {
                $self->res->code(403);
                $self->render( text => 'no INBOX directory');
                return;
            }
            $self->render_later;
        })
    }
    else {
        # / (upload page)
        $r->get('/' => 'home');
        $b = $r->under('/:user' => sub {
            my $self = shift;
            if (not $self->param('user') =~ /^(\w+)$/){
               $self->res->code(403);
               $self->render( text => 'invalid user');
               return;
            }
            my $username = $1;
            $self->stash('user'=>$username);
            my $root_dir = $ENV{US_ROOT} || '/tmp';
            my $root = $root_dir . '/'. $username . '/INBOX';

            $self->stash(root=>$root);

            my $uid = getpwnam($username);
            if (not $uid){
                $self->res->code(403);
                $self->render( text => 'unknown user');
                return;
            }
            # lets see if  we can do a bit of user switching
            if ($< == 0 and $> != $uid ){
                # switch uid
                $> = 0;
                $> = $uid;
                if ($> != $uid){
                    $self->res->code(403);
                    $self->render( text => 'access denied: '.$!);
                    return;
                }
            }

            if ( -l $root or not -w $root or not -d $root ) {
                $self->res->code(403);
                $self->render( text => 'no INBOX directory in sandbox');
                return;
            }
            $self->render_later;
        });

    }
    if ($ENV{US_TAGMODE}){        # / (upload page)
        if ($ENV{US_SINGLEUSER}){
            $r->get('/' => 'home');
        }
        $b = $b->under('/:tag' => sub {
            my $self = shift;
            if (not $self->param('tag') =~ /^([a-z0-9]+)$/){
               $self->res->code(403);
               $self->render( text => 'tags must consist of 0-9 and a-z');
               return;
            }
            my $tag = $1;
            $self->stash('tag'=>$tag);
            $self->render_later;
        });
    }
    $a = $b->under(sub {
        my $self = shift;
        my $sessionkey = $self->session('skey');
        if (not $sessionkey){
            my $newKey = hmac_sha1_sum(rand,time);
            $self->session(skey => $newKey);
        }
        if (not $self->session('skey') =~ /^(\w+)$/){
            $self->res->code(403);
            $self->render( text => 'invalid session key');
            return;
        }
        $self->stash(skey => $1);
        $self->render_later;
    });

    # / (upload page)
    $a->get('/' => sub {
        my $self = shift;
        if ($self->req->url ne '' and $self->req->url !~ m{/$}){
            $self->redirect_to($self->req->url.'/');
        }
        else {
            $self->render(template=>'resumable');
        }
    });

    # GET /upload (retrieves stored file list)
    $a->get('/upload' => sub {
        my $self = shift;
        my $root = $self->stash('root');
        my $sessionkey = $self->stash('skey');
        my $name = '.'.$sessionkey.'-'.cleanName($self->param('resumableIdentifier'));
        my $chunkNr = int($self->param('resumableChunkNumber'));
        if (-e $root.'/'.$name.'.'.$chunkNr){
            $self->render(text=>'chunk ok',status=>200);
        }
        else {
            $self->render(text=>'chunk missing',status=>206);
        }
    });

    # POST /upload (push one or more files to app)
    $a->post('/upload' => sub {
        my $self    = shift;
        my $sessionkey = $self->stash('skey');
        my $root = $self->stash('root');
        my $name = '.'.$sessionkey.'-'.cleanName($self->param('resumableIdentifier'));
        my $chunkNr = int($self->param('resumableChunkNumber'));
        my $chunkSize = int($self->param('resumableChunkSize'));
        my $chunkTotal = int($self->param('resumableTotalChunks'));
        my $thisChunkSize = int($self->param('resumableCurrentChunkSize'));
        my $data = $self->req->upload('file')->slurp;
        my $dataLen = length($data);
        if ($dataLen != $thisChunkSize){
            $self->render(status=>206,text=>'upload size was not as expected');
            return;
        }

        eval {
            use autodie;
            local $SIG{__DIE__};
            my $file = $root.'/'.$name;
            if (not -e $file){
                open my $c,'>>',$file;
                close $c
            }
            open my $fh, "+<", $file;
            binmode $fh,':raw';
            sysseek $fh,($chunkNr-1)*$chunkSize,SEEK_SET;
            syswrite $fh,$data;
            close $fh;
            open my $touch, '>',$file.'.'.$chunkNr;
            close $touch;
            $self->render(text=>'ok',status=>200);
            my @rm;
            for (1..$chunkTotal){
                my $chunk = $root.'/'.$name.'.'.$_;
                return if not -e $chunk;
                push @rm, $chunk;
            }
            my $tag = $self->stash('tag') ? $self->stash('tag').'-' : '';
            rename $file,$root.'/'.$tag.strftime('%Y-%m-%d_%H-%M-%S_',localtime(time)).$self->param('resumableFilename');
            unlink @rm;
        };

        if ($@ and $@->isa('autodie::exception')) {
            $self->app->log->warn($@->errno);
            $self->render(status=>404,text=>$@->errno);
        } elsif ($@) {
            die $@;
        }
    });
}

sub cleanName {
    my $name = shift;
    $name =~ s/[^-_a-z0-9]+/_/ig;
    return $name;
}

1;
